### ISSUE TITLE
DockerHost startup hung on apt-get install, in BlueBox with Ubuntu 14.04 

### ISSUE THREAD

**Issue by aledsage** [brooklyncentral/clocker#249]
When attempting to deploy clocker+calico to BlueBox 2.0, the `DockerHost` task `ssh: installingkernel DockerHostImpl{id=IEC79bVJ}` hung.

This was on Ubuntu 14.04, on a m1.large instance.

When I ssh'ed to the box, I could see that the `apt-get` command was still running (a day later):

    $ ps aux | grep apt-get
    root      1704  0.0  0.0  63668  2092 ?        S    Mar09   0:00 sudo -E -n -S -- apt-get install -y --allow-unauthenticated software-properties-common linux-generic-lts-vivid
    root      1705  3.7  0.4  73780 34940 ?        S    Mar09  35:43 apt-get install -y --allow-unauthenticated software-properties-common linux-generic-lts-vivid

It seems that apt believes the install had completed:

    $ apt --installed list | grep vivid
    
    WARNING: apt does not have a stable CLI interface yet. Use with caution in scripts.
    
    linux-generic-lts-vivid/trusty-updates,trusty-security,now 3.19.0.51.36 amd64 [installed]
    linux-headers-generic-lts-vivid/trusty-updates,trusty-security,now 3.19.0.51.36 amd64 [installed]
    linux-image-generic-lts-vivid/trusty-updates,trusty-security,now 3.19.0.51.36 amd64 [installed]

There wasn't anything interested in the logs that I could find.

`/var/log/apt/term.log` contained:

    Log started: 2016-03-09  22:53:09

`/var/log/apt/history.log` contained:

    Start-Date: 2016-03-09  22:53:09
    Commandline: apt-get install -y --allow-unauthenticated software-properties-common linux-generic-lts-vivid
    Install: libnl-genl-3-200:amd64 (3.2.21-1, automatic), linux-image-extra-3.19.0-51-generic:amd64 (3.19.0-51.58~14.04.1, automatic), linux-image-3.19.0-51-generic:amd64 (3.19.0-51.58~14.04.1, automatic), iw:amd64 (3.4-1, automatic), linux-headers-3.19.0-51-generic:amd64 (3.19.0-51.58~14.04.1, automatic), wireless-regdb:amd64 (2013.02.13-1ubuntu1, automatic), linux-image-generic-lts-vivid:amd64 (3.19.0.51.36, automatic), thermald:amd64 (1.4.3-5~14.04.2, automatic), libnl-3-200:amd64 (3.2.21-1, automatic), linux-headers-generic-lts-vivid:amd64 (3.19.0.51.36, automatic), linux-generic-lts-vivid:amd64 (3.19.0.51.36), crda:amd64 (1.1.2-1ubuntu2, automatic), linux-firmware:amd64 (1.127.20, automatic), linux-headers-3.19.0-51:amd64 (3.19.0-51.58~14.04.1, automatic)

The stderr of the task is shown below (note the "Segmentation fault", which happened on my two tests in the two different projects - and the repetition after the Segmentation fault):

```
sudo: unable to resolve host brooklyn-o3snzg-amp-master-clocke-gn3h-docker-host-iec7-bb9
sudo: unable to resolve host brooklyn-o3snzg-amp-master-clocke-gn3h-docker-host-iec7-bb9
Done.
Running depmod.
update-initramfs: deferring update (hook will be called later)
Examining /etc/kernel/postinst.d.
run-parts: executing /etc/kernel/postinst.d/apt-auto-removal 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
run-parts: executing /etc/kernel/postinst.d/initramfs-tools 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
update-initramfs: Generating /boot/initrd.img-3.19.0-51-generic
run-parts: executing /etc/kernel/postinst.d/update-notifier 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
run-parts: executing /etc/kernel/postinst.d/x-grub-legacy-ec2 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
Searching for GRUB installation directory ... found: /boot/grub
Searching for default file ... found: /boot/grub/default
Testing for an existing GRUB menu.lst file ... found: /boot/grub/menu.lst
Searching for splash image ... none found, skipping ...
Found kernel: /boot/vmlinuz-3.13.0-79-generic
Found kernel: /boot/vmlinuz-3.19.0-51-generic
Found kernel: /boot/vmlinuz-3.13.0-79-generic
Replacing config file /run/grub/menu.lst with new version
Updating /boot/grub/menu.lst ... done
run-parts: executing /etc/kernel/postinst.d/zz-update-grub 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-3.19.0-51-generic
Found initrd image: /boot/initrd.img-3.19.0-51-generic
Found linux image: /boot/vmlinuz-3.13.0-79-generic
Found initrd image: /boot/initrd.img-3.13.0-79-generic
Segmentation fault
done
run-parts: executing /etc/kernel/postinst.d/apt-auto-removal 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
run-parts: executing /etc/kernel/postinst.d/initramfs-tools 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
update-initramfs: Generating /boot/initrd.img-3.19.0-51-generic
run-parts: executing /etc/kernel/postinst.d/update-notifier 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
run-parts: executing /etc/kernel/postinst.d/x-grub-legacy-ec2 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
Searching for GRUB installation directory ... found: /boot/grub
Searching for default file ... found: /boot/grub/default
Testing for an existing GRUB menu.lst file ... found: /boot/grub/menu.lst
Searching for splash image ... none found, skipping ...
Found kernel: /boot/vmlinuz-3.19.0-51-generic
Found kernel: /boot/vmlinuz-3.13.0-79-generic
Updating /boot/grub/menu.lst ... done
run-parts: executing /etc/kernel/postinst.d/zz-update-grub 3.19.0-51-generic /boot/vmlinuz-3.19.0-51-generic
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-3.19.0-51-generic
Found initrd image: /boot/initrd.img-3.19.0-51-generic
Found linux image: /boot/vmlinuz-3.13.0-79-generic
Found initrd image: /boot/initrd.img-3.13.0-79-generic
```

The stdout was:

```
/usr/bin/apt-get
apt-get exists, doing update
Get:1 http://security.ubuntu.com trusty-security InRelease [65.9 kB]
Ign http://nova.clouds.archive.ubuntu.com trusty InRelease
Get:2 http://security.ubuntu.com trusty-security/main Sources [106 kB]
Get:3 http://nova.clouds.archive.ubuntu.com trusty-updates InRelease [65.9 kB]
Get:4 http://security.ubuntu.com trusty-security/universe Sources [33.3 kB]
Get:5 http://security.ubuntu.com trusty-security/main amd64 Packages [431 kB]
Get:6 http://nova.clouds.archive.ubuntu.com trusty-backports InRelease [65.9 kB]
Get:7 http://security.ubuntu.com trusty-security/universe amd64 Packages [124 kB]
Get:8 http://security.ubuntu.com trusty-security/main Translation-en [236 kB]
Hit http://nova.clouds.archive.ubuntu.com trusty Release.gpg
Get:9 http://security.ubuntu.com trusty-security/universe Translation-en [73.1 kB]
Get:10 http://nova.clouds.archive.ubuntu.com trusty-updates/main Sources [262 kB]
Get:11 http://nova.clouds.archive.ubuntu.com trusty-updates/restricted Sources [5,352 B]
Get:12 http://nova.clouds.archive.ubuntu.com trusty-updates/universe Sources [150 kB]
Get:13 http://nova.clouds.archive.ubuntu.com trusty-updates/multiverse Sources [5,943 B]
Get:14 http://nova.clouds.archive.ubuntu.com trusty-updates/main amd64 Packages [714 kB]
Get:15 http://nova.clouds.archive.ubuntu.com trusty-updates/restricted amd64 Packages [15.9 kB]
Get:16 http://nova.clouds.archive.ubuntu.com trusty-updates/universe amd64 Packages [339 kB]
Get:17 http://nova.clouds.archive.ubuntu.com trusty-updates/multiverse amd64 Packages [13.2 kB]
Get:18 http://nova.clouds.archive.ubuntu.com trusty-updates/main Translation-en [361 kB]
Get:19 http://nova.clouds.archive.ubuntu.com trusty-updates/multiverse Translation-en [7,227 B]
Get:20 http://nova.clouds.archive.ubuntu.com trusty-updates/restricted Translation-en [3,699 B]
Get:21 http://nova.clouds.archive.ubuntu.com trusty-updates/universe Translation-en [181 kB]
Get:22 http://nova.clouds.archive.ubuntu.com trusty-backports/main Sources [8,661 B]
Get:23 http://nova.clouds.archive.ubuntu.com trusty-backports/restricted Sources [28 B]
Get:24 http://nova.clouds.archive.ubuntu.com trusty-backports/universe Sources [33.2 kB]
Get:25 http://nova.clouds.archive.ubuntu.com trusty-backports/multiverse Sources [1,898 B]
Get:26 http://nova.clouds.archive.ubuntu.com trusty-backports/main amd64 Packages [9,787 B]
Get:27 http://nova.clouds.archive.ubuntu.com trusty-backports/restricted amd64 Packages [28 B]
Get:28 http://nova.clouds.archive.ubuntu.com trusty-backports/universe amd64 Packages [40.0 kB]
Get:29 http://nova.clouds.archive.ubuntu.com trusty-backports/multiverse amd64 Packages [1,571 B]
Get:30 http://nova.clouds.archive.ubuntu.com trusty-backports/main Translation-en [5,843 B]
Get:31 http://nova.clouds.archive.ubuntu.com trusty-backports/multiverse Translation-en [1,215 B]
Get:32 http://nova.clouds.archive.ubuntu.com trusty-backports/restricted Translation-en [28 B]
Get:33 http://nova.clouds.archive.ubuntu.com trusty-backports/universe Translation-en [34.7 kB]
Hit http://nova.clouds.archive.ubuntu.com trusty Release
Get:34 http://nova.clouds.archive.ubuntu.com trusty/main Sources [1,064 kB]
Get:35 http://nova.clouds.archive.ubuntu.com trusty/restricted Sources [5,433 B]
Get:36 http://nova.clouds.archive.ubuntu.com trusty/universe Sources [6,399 kB]
Get:37 http://nova.clouds.archive.ubuntu.com trusty/multiverse Sources [174 kB]
Hit http://nova.clouds.archive.ubuntu.com trusty/main amd64 Packages
Hit http://nova.clouds.archive.ubuntu.com trusty/restricted amd64 Packages
Hit http://nova.clouds.archive.ubuntu.com trusty/universe amd64 Packages
Hit http://nova.clouds.archive.ubuntu.com trusty/multiverse amd64 Packages
Hit http://nova.clouds.archive.ubuntu.com trusty/main Translation-en
Hit http://nova.clouds.archive.ubuntu.com trusty/multiverse Translation-en
Hit http://nova.clouds.archive.ubuntu.com trusty/restricted Translation-en
Hit http://nova.clouds.archive.ubuntu.com trusty/universe Translation-en
Ign http://nova.clouds.archive.ubuntu.com trusty/main Translation-en_US
Ign http://nova.clouds.archive.ubuntu.com trusty/multiverse Translation-en_US
Ign http://nova.clouds.archive.ubuntu.com trusty/restricted Translation-en_US
Ign http://nova.clouds.archive.ubuntu.com trusty/universe Translation-en_US
Fetched 11.0 MB in 15s (726 kB/s)
Reading package lists...
Reading package lists...
Building dependency tree...
Reading state information...
software-properties-common is already the newest version.
The following extra packages will be installed:
  crda iw libnl-3-200 libnl-genl-3-200 linux-firmware linux-headers-3.19.0-51
  linux-headers-3.19.0-51-generic linux-headers-generic-lts-vivid
  linux-image-3.19.0-51-generic linux-image-extra-3.19.0-51-generic
  linux-image-generic-lts-vivid thermald wireless-regdb
Suggested packages:
  fdutils linux-lts-vivid-tools
The following NEW packages will be installed:
  crda iw libnl-3-200 libnl-genl-3-200 linux-firmware linux-generic-lts-vivid
  linux-headers-3.19.0-51 linux-headers-3.19.0-51-generic
  linux-headers-generic-lts-vivid linux-image-3.19.0-51-generic
  linux-image-extra-3.19.0-51-generic linux-image-generic-lts-vivid thermald
  wireless-regdb
0 upgraded, 14 newly installed, 0 to remove and 21 not upgraded.
Need to get 89.9 MB of archives.
After this operation, 375 MB of additional disk space will be used.
Get:1 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty/main libnl-3-200 amd64 3.2.21-1 [44.5 kB]
Get:2 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty/main libnl-genl-3-200 amd64 3.2.21-1 [10.2 kB]
Get:3 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main linux-image-3.19.0-51-generic amd64 3.19.0-51.58~14.04.1 [16.7 MB]
Get:4 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty/main wireless-regdb all 2013.02.13-1ubuntu1 [6,456 B]
Get:5 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty/main crda amd64 1.1.2-1ubuntu2 [15.2 kB]
Get:6 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty/main iw amd64 3.4-1 [51.7 kB]
Get:7 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main linux-firmware all 1.127.20 [24.5 MB]
Get:8 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main linux-image-extra-3.19.0-51-generic amd64 3.19.0-51.58~14.04.1 [38.3 MB]
Get:9 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main linux-image-generic-lts-vivid amd64 3.19.0.51.36 [2,270 B]
Get:10 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main linux-headers-3.19.0-51 all 3.19.0-51.58~14.04.1 [9,299 kB]
Get:11 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main linux-headers-3.19.0-51-generic amd64 3.19.0-51.58~14.04.1 [744 kB]
Get:12 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main linux-headers-generic-lts-vivid amd64 3.19.0.51.36 [2,248 B]
Get:13 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main linux-generic-lts-vivid amd64 3.19.0.51.36 [1,798 B]
Get:14 http://nova.clouds.archive.ubuntu.com/ubuntu/ trusty-updates/main thermald amd64 1.4.3-5~14.04.2 [202 kB]
Fetched 89.9 MB in 9s (9,770 kB/s)
Selecting previously unselected package libnl-3-200:amd64.
(Reading database ... 51154 files and directories currently installed.)
Preparing to unpack .../libnl-3-200_3.2.21-1_amd64.deb ...
Unpacking libnl-3-200:amd64 (3.2.21-1) ...
Selecting previously unselected package libnl-genl-3-200:amd64.
Preparing to unpack .../libnl-genl-3-200_3.2.21-1_amd64.deb ...
Unpacking libnl-genl-3-200:amd64 (3.2.21-1) ...
Selecting previously unselected package linux-image-3.19.0-51-generic.
Preparing to unpack .../linux-image-3.19.0-51-generic_3.19.0-51.58~14.04.1_amd64.deb ...
Unpacking linux-image-3.19.0-51-generic (3.19.0-51.58~14.04.1) ...
Selecting previously unselected package wireless-regdb.
Preparing to unpack .../wireless-regdb_2013.02.13-1ubuntu1_all.deb ...
Unpacking wireless-regdb (2013.02.13-1ubuntu1) ...
Selecting previously unselected package crda.
Preparing to unpack .../crda_1.1.2-1ubuntu2_amd64.deb ...
Unpacking crda (1.1.2-1ubuntu2) ...
Selecting previously unselected package iw.
Preparing to unpack .../archives/iw_3.4-1_amd64.deb ...
Unpacking iw (3.4-1) ...
Selecting previously unselected package linux-firmware.
Preparing to unpack .../linux-firmware_1.127.20_all.deb ...
Unpacking linux-firmware (1.127.20) ...
Selecting previously unselected package linux-image-extra-3.19.0-51-generic.
Preparing to unpack .../linux-image-extra-3.19.0-51-generic_3.19.0-51.58~14.04.1_amd64.deb ...
Unpacking linux-image-extra-3.19.0-51-generic (3.19.0-51.58~14.04.1) ...
Selecting previously unselected package linux-image-generic-lts-vivid.
Preparing to unpack .../linux-image-generic-lts-vivid_3.19.0.51.36_amd64.deb ...
Unpacking linux-image-generic-lts-vivid (3.19.0.51.36) ...
Selecting previously unselected package linux-headers-3.19.0-51.
Preparing to unpack .../linux-headers-3.19.0-51_3.19.0-51.58~14.04.1_all.deb ...
Unpacking linux-headers-3.19.0-51 (3.19.0-51.58~14.04.1) ...
Selecting previously unselected package linux-headers-3.19.0-51-generic.
Preparing to unpack .../linux-headers-3.19.0-51-generic_3.19.0-51.58~14.04.1_amd64.deb ...
Unpacking linux-headers-3.19.0-51-generic (3.19.0-51.58~14.04.1) ...
Selecting previously unselected package linux-headers-generic-lts-vivid.
Preparing to unpack .../linux-headers-generic-lts-vivid_3.19.0.51.36_amd64.deb ...
Unpacking linux-headers-generic-lts-vivid (3.19.0.51.36) ...
Selecting previously unselected package linux-generic-lts-vivid.
Preparing to unpack .../linux-generic-lts-vivid_3.19.0.51.36_amd64.deb ...
Unpacking linux-generic-lts-vivid (3.19.0.51.36) ...
Selecting previously unselected package thermald.
Preparing to unpack .../thermald_1.4.3-5~14.04.2_amd64.deb ...
Unpacking thermald (1.4.3-5~14.04.2) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Processing triggers for ureadahead (0.100.0-16) ...
Setting up libnl-3-200:amd64 (3.2.21-1) ...
Setting up libnl-genl-3-200:amd64 (3.2.21-1) ...
Setting up linux-image-3.19.0-51-generic (3.19.0-51.58~14.04.1) ...
Setting up wireless-regdb (2013.02.13-1ubuntu1) ...
Setting up crda (1.1.2-1ubuntu2) ...
Setting up iw (3.4-1) ...
Setting up linux-firmware (1.127.20) ...
Setting up linux-image-extra-3.19.0-51-generic (3.19.0-51.58~14.04.1) ...
```

The stdin was:

```
( { which zypper && { echo zypper exists, doing refresh && (( if test "$UID" -eq 0; then ( zypper --non-interactive --no-gpg-checks refresh ); else sudo -E -n -S -- zypper --non-interactive --no-gpg-checks refresh; fi ) || true) && ( if test "$UID" -eq 0; then ( zypper --non-interactive --no-gpg-checks install software-properties-common linux-generic-lts-vivid ); else sudo -E -n -S -- zypper --non-interactive --no-gpg-checks install software-properties-common linux-generic-lts-vivid; fi ) ; } ; } || { which apt-get && { echo apt-get exists, doing update && export DEBIAN_FRONTEND=noninteractive && (( if test "$UID" -eq 0; then ( apt-get update ); else sudo -E -n -S -- apt-get update; fi ) || true) && ( if test "$UID" -eq 0; then ( apt-get install -y --allow-unauthenticated software-properties-common linux-generic-lts-vivid ); else sudo -E -n -S -- apt-get install -y --allow-unauthenticated software-properties-common linux-generic-lts-vivid; fi ) ; } ; } || { which yum && { echo yum exists, doing update && (( if test "$UID" -eq 0; then ( yum check-update ); else sudo -E -n -S -- yum check-update; fi ) || true) && (( if test "$UID" -eq 0; then ( yum -y install epel-release ); else sudo -E -n -S -- yum -y install epel-release; fi ) || true) && (( if test "$UID" -eq 0; then ( yum upgrade -y ca-certificates --disablerepo=epel ); else sudo -E -n -S -- yum upgrade -y ca-certificates --disablerepo=epel; fi ) || true) && ( if test "$UID" -eq 0; then ( yum -y --nogpgcheck install software-properties-common linux-generic-lts-vivid ); else sudo -E -n -S -- yum -y --nogpgcheck install software-properties-common linux-generic-lts-vivid; fi ) ; } ; } || { which brew && brew install software-properties-common linux-generic-lts-vivid ; } || { which port && ( if test "$UID" -eq 0; then ( port install software-properties-common linux-generic-lts-vivid ); else sudo -E -n -S -- port install software-properties-common linux-generic-lts-vivid; fi ) ; } || (( echo "WARNING: no known/successful package manager to install software-properties-common linux-generic-lts-vivid, may fail subsequently" | tee /dev/stderr ) || true) )
( if test "$UID" -eq 0; then ( reboot ); else sudo -E -n -S -- reboot; fi )
```

**Comment by aledsage** [comment]
@grkvlt  When I ran this on `Ubuntu 15.04`, it worked (presumably because the kernel update is not required). However, I was surprised by the logging. The host's install task included a subtask for `installingkernel`, which tried to execute `apt-get install -y --allow-unauthenticated software-properties-common linux-generic-lts-vivid`. This gave a non-zero exit code (and the stderr `Unable to locate package linux-generic-lts-vivid`). Because of that exit code, we got the warning `no known/successful package manager to install software-properties-common linux-generic-lts-vivid, may fail subsequently`. But it continued, and succeeded.

I'd have expected it to not try to upgrade the kernel if it already has the right version.
